libraries/AP_InertialSensor/AP_InertialSensor_L3G4200D.cpp

```diff
@@ -243,41 +300,51 @@ void AP_InertialSensor_L3G4200D::_accumulate(void)
         num_samples_available = fifo_status & L3G4200D_REG_FIFO_SRC_ENTRIES_MASK;
     }
 
+    // read the samples and apply the filter
     if (num_samples_available > 0) {
         // read all the entries in one go, using AUTO_INCREMENT. This saves a lot of time on I2C setup
         int16_t buffer[num_samples_available][3];
-        if (!_dev->read_registers(L3G4200D_REG_XL | L3G4200D_REG_AUTO_INCREMENT,
+        if (_dev_gyro->read_registers(L3G4200D_REG_XL | L3G4200D_REG_AUTO_INCREMENT,
                                   (uint8_t *)&buffer, sizeof(buffer))) {
             for (uint8_t i=0; i < num_samples_available; i++) {
                 Vector3f gyro = Vector3f(buffer[i][0], -buffer[i][1], -buffer[i][2]);
                 // Adjust for chip scaling to get radians/sec
+                //hal.console->printf("gyro %f \r\n",gyro.x); 
                 gyro *= L3G4200D_GYRO_SCALE_R_S;
                 _rotate_and_correct_gyro(_gyro_instance, gyro);
                 _notify_new_gyro_raw_sample(_gyro_instance, gyro);
             }
         }
     }
+}
+
+void AP_InertialSensor_L3G4200D::_accumulate_accel (void)
+{
+    uint8_t num_samples_available;
+    uint8_t fifo_status = 0;
 
     // Read accelerometer FIFO to find out how many samples are available
-    _dev->read_registers(ADXL345_ACCELEROMETER_ADXLREG_FIFO_STATUS,
+    _dev_accel->read_registers(ADXL345_ACCELEROMETER_ADXLREG_FIFO_STATUS,
                          &fifo_status, 1);
     num_samples_available = fifo_status & 0x3F;
 
     // read the samples and apply the filter
     if (num_samples_available > 0) {
         int16_t buffer[num_samples_available][3];
-        if (!_dev->read_registers_multiple(ADXL345_ACCELEROMETER_ADXLREG_DATAX0,
-                                           (uint8_t *)buffer, sizeof(buffer[0]),
-                                           num_samples_available)) {
-            for (uint8_t i=0; i<num_samples_available; i++) {
+        for (uint8_t i=0; i<num_samples_available; i++) 
```

Comment:
what's the problem with read_registers_multiple() here?

Reply:
**read_registers_multiple** is not working. I have just retested again before writing this feedback.

The other weird thing is that it is used here as:
```
 if (!_dev->read_registers_multiple(ADXL345_ACCELEROMETER_ADXLREG_DATAX0,
                                           (uint8_t *)buffer, sizeof(buffer[0]),
                                           num_samples_available)) {
}
```
 the **!** is used while it is supposed to return true not false.
also this is the only place in code where I found this function call and it did not work correctly when I tried it. 

Maybe the function itself needs fixes or not supported on all machines. I am using raspberry zero.

Reply:
@tridge would you please advise about read_registers_multiple call.

Reply:
humn... this function might be wrong as there are no users left and chibios hal doesn't even implement it.

Reply:
I believe so, and the alternative function that I use is working properly with me.